docs/getting-started/walkthroughs/datastream_api.md

```diff
@@ -0,0 +1,925 @@
+        // Get the current state for the current key
+        Boolean lastTransactionWasSmall = flagState.value();
+
+        // Check if the flag is set
+        if (lastTransactionWasSmall != null) {
+            if (transaction.getAmount() > LARGE_AMOUNT) {
+                //Output an alert downstream
+                Alert alert = new Alert();
+                alert.setId(transaction.getAccountId());
+
+                collector.collect(alert);            
+            }
+
+            // Clean up our state
+            flagState.clear();
+        }
+
+        if (transaction.getAmount() < SMALL_AMOUNT) {
+            // set the flag to true
+            flagState.update(true);
+        }
+    }
+}
+{% endhighlight %}
+</div>
+
+<div data-lang="scala" markdown="1">
+{% highlight scala %}
+    override def processElement(
+            transaction: Transaction,
+            context: Context,
+            collector: Collector[Alert]): Unit = {
+
+        // Get the current state for the current key
+        val lastTransactionWasSmall = flagState.value
+
+        // Check if the flag is set
+        if (lastTransactionWasSmall != null) {
+            if (transaction.getAmount() > LARGE_AMOUNT) {
+                //Output an alert downstream
+                Alert alert = new Alert
+                alert.setId(transaction.getAccountId)
+
+                collector.collect(alert)
+            }
+
+            // Clean up our state
+            flagState.clear()
+        }
+
+        if (transaction.getAmount() < SMALL_AMOUNT) {
+            // set the flag to true
+            flagState.update(true)
+        }
+    }
+}
+{% endhighlight %}
+</div>
+</div>
+
+For every transaction, the fraud detector checks the state of the flag for that account.
+Remember, `ValueState` is always scoped to the current key.
+If the flag is non-null, then the last transaction seen for that key was small, and so if the amount for this transaction is large, then the detector outputs a fraud alert.
+
+After that check, the flag state is unconditionally cleared.
+Either the current transaction caused a fraud alert, and the pattern is over, or the current transaction did not cause an alert, and the pattern is broken and needs to be restarted.
+
+Finally, the transaction amount is checked to see if it is small.
+If so, then the flag is set so that it can be checked by the next event.
+Notice that `ValueState<Boolean>` actually has three states, unset ( `null`), `true`, and `false`, because all `ValueState`'s are nullable.
+This job only makes use of unset ( `null`) and `true` to check whether the flag is set or not.
+
+## Fraud Detector v2: State + Time = &#10084;&#65039;
+
+Many event-driven applications require a strong notion of time to complement their state.
+For example, suppose you wanted to set a 24-hour timeout to your fraud detector; i.e., in the previous example transactions 3 and 4 would only be considered fraud if they occurred within 24 hours of each other.
```

Comment:
Relate the requirement to the real world by explaining that scammers don't wait to long to reduce the chance that the test transaction is noticed?

Reply:
Not sure if it makes sense, but maybe reduce the time out to 30 secs or a minute? 24h is a long time for an example. Could be argued that if two such different transactions are done is such a short time it's a clear sign for fraud and mention that this pattern is not designed to catch all fraudulent transactions.